src/main/java/com/googlesource/gerrit/plugins/oauth/GitLabApi.java

```diff
@@ -0,0 +1,148 @@
+package com.googlesource.gerrit.plugins.oauth;
+
+
+import com.google.common.io.BaseEncoding;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import org.scribe.builder.api.DefaultApi20;
+import org.scribe.exceptions.OAuthException;
+import org.scribe.extractors.AccessTokenExtractor;
+import org.scribe.model.OAuthConfig;
+import org.scribe.model.OAuthRequest;
+import org.scribe.model.Response;
+import org.scribe.model.Token;
+import org.scribe.model.Verb;
+import org.scribe.model.Verifier;
+import org.scribe.oauth.OAuthService;
+
+import static com.google.gerrit.server.OutputFormat.JSON;
+import static java.lang.String.format;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+import static org.scribe.model.OAuthConstants.ACCESS_TOKEN;
+import static org.scribe.model.OAuthConstants.CODE;
+
+public class GitLabApi extends DefaultApi20{
+    private static final String AUTHORIZE_URL =
+            "%s/oauth/authorize?client_id=%s&response_type=code&redirect_uri=%s";
+
+    private final String rootUrl;
+
+    public GitLabApi(String rootUrl) {
+        this.rootUrl = rootUrl;
+    }
+
+    @Override
+    public String getAuthorizationUrl(OAuthConfig config) {
+        return String.format(AUTHORIZE_URL, rootUrl, config.getApiKey(),config.getCallback());
+    }
+
+    @Override
+    public String getAccessTokenEndpoint() {
+        return String.format("%s/oauth/token", rootUrl);
+    }
+
+    @Override
+    public Verb getAccessTokenVerb() {
+        return Verb.POST;
+    }
+
+    @Override
+    public OAuthService createService(OAuthConfig config) {
+        return new GitLabOAuthService(this, config);
+    }
+
+    @Override
+    public AccessTokenExtractor getAccessTokenExtractor() {
+        return new GitLabTokenExtractor();
+    }
+
+    private static final class GitLabOAuthService implements OAuthService {
```

Comment:
It's because of GitLab request specialty(https://docs.gitlab.com/ce/api/oauth2.html), the solution with creating new essences outside the class doesn't look well for understanding.